sql/core/src/test/scala/org/apache/spark/sql/streaming/TransformWithStateInitialStateSuite.scala

```diff
@@ -83,6 +98,49 @@ abstract class StatefulProcessorWithInitialStateTestClass[V]
   }
 }
 
+/**
+ * Stateful processor that will take a union dataframe output from state data source reader,
+ * with flattened state data source rows.
+ */
+class InitialStatefulProcessorWithStateDataSource
+  extends StatefulProcessorWithInitialStateTestClass[UnionInitialStateRow] {
+  override def handleInitialState(
+      key: String, initialState: UnionInitialStateRow, timerValues: TimerValues): Unit = {
+    if (initialState.value.isDefined) {
+      _valState.update(initialState.value.get.toDouble)
+    } else if (initialState.listValue.isDefined) {
+      _listState.appendValue(initialState.listValue.get.toDouble)
+    } else if (initialState.userMapKey.isDefined) {
+      _mapState.updateValue(
+        (initialState.userMapKey.get.charAt(0) - 'a').toDouble,
+        initialState.userMapValue.get.toInt)
+    }
+  }
+}
+
+/**
+ * Stateful processor that will take a union dataframe output from state data source reader,
+ * with composite type state data source rows.
```

Comment:
ditto

Reply:
Done. Modified handle initial state to be simply put the values in the initialState dataframe into state variables.